swayfmt/src/formatter/shape.rs

```diff
@@ -188,6 +194,18 @@ impl Shape {
     // |<---->|          indent
     //        |<--->|    width
     //
+    /// Create a copy of
+    pub(crate) fn from(shape: &Shape, width: Option<usize>, code_line: Option<CodeLine>) -> Self {
```

Comment:
One alternative to using multiple `Option`s in a constructor's set of arguments (which can be a little tricky to read on the calling side if you don't know what the arguments happen to refer to) is to use the "builder method" pattern. E.g.

```rust
impl Shape {
    // ...

    pub(crate) fn width(mut self, w: usize) -> Self {
        self.update_width(w);
        self
    }

    pub(crate) fn line(mut self, code_line: CodeLinde) -> Self {
        self.update_line_settings(code_line);
        self
    }

    // ...
}
```

This way, rather than constructing a new Shape like so:

```rust
let new_shape = Shape::from(&formatter.shape, Some(new_width) Some(new_line));
```

user's could construct new `Shape` instances like this:

```rust
let new_shape = formatter.shape.clone().width(new_width).line(new_line);
```
This approach is particularly useful in the case that you don't want to update *all* of the options at once, e.g. here we only update the width:
```rust
let new_shape = formatter.shape.clone().width(new_width);
```
whereas the current approach would require specifying a redundant `None`, e.g.
```rust
let new_shape = Shape::new(&formatter.shape, Some(new_width), None);
```

You start to notice the benefits of the builder pattern more when you have many more optional parameters (e.g. imagine if we had 5 optional parameters but only wanted to specify `width`).

### Sidenotes

Some folks prefer the `with_` prefix for builder methods, e.g. `shape.with_width(w).with_line(l)`, but I personally don't mind either approach, as long as they're used consistently and don't cause ambiguity with any of the other methods. 

There are also other variations on the builder method where a dedicated `ShapeBuilder` type would be used with a `.build()` call chained on at the end to construct the actual `Shape`, however this approach is more useful when building much larger sized types that store a lot of data.

You can read more about the builder pattern [here](https://doc.rust-lang.org/1.0.0/style/ownership/builders.html).

Reply:
Ah much better 😄 Thank you! I'll add these changes in!

Reply:
I think I found a decent compromise to this, let me know if it needs any other changes!